### DISCUSSION TITLE
File name column takes to much space (even truncated with custom code).)

### DISCUSSION THREAD

**Discussion by gety9** [minad/vertico#438] [Q&A]
Title. This issue was discussed before. I applied suggested custom code. It visually truncated length of line name, but file name column still ocupies lots of space.

Screen shots
Buffers section has correct column width
![Screenshot_20240204-111814](https://github.com/minad/vertico/assets/16415882/9242b9cf-b772-4566-927d-01a06d188961)

But when you scroll to files it worsens
![Screenshot_20240204-111826](https://github.com/minad/vertico/assets/16415882/c0961814-b950-4c7c-a5ef-37fc85ddbd7a)

And at bottom (files section) hardly visible
![Screenshot_20240204-111840](https://github.com/minad/vertico/assets/16415882/e80be775-fe7e-476d-8dc4-7e5490c2769e)

Config
```
(use-package consult
  :config
  (consult-customize
    consult-buffer :preview-key ">"
    consult-line   :preview-key ">"))

(use-package embark
  :demand t)
(use-package embark-consult)

(defun my/vertico-truncate-candidates (args)
  (if-let ((arg (car args))
           (type (get-text-property 0 'multi-category arg))
           ((eq (car-safe type) 'file))
           (w (max 25 (- (window-width) 55)))
           (l (length arg))
           ((> l w)))
      (setcar args (concat "…" (truncate-string-to-width arg l (- l w)))))
  args)
(advice-add #'vertico--format-candidate :filter-args #'my/vertico-truncate-candidates)
(use-package vertico
  :init
  (vertico-mode))

(use-package marginalia
  :init
  (marginalia-mode))

(use-package orderless
  :demand t
  :init
  (setq completion-styles '(orderless)
    read-file-name-completion-ignore-case t
    read-buffer-completion-ignore-case t
    completion-ignore-case t))�
```

**Comment by minad** [comment]
Please don't open new issues if the issue has been discussed before. Please comment instead on existing issues or discussions. You can give https://github.com/jdtsmith/vertico-truncate a try. Vertico does not truncate candidates, since candidates are considered more important than annotations.

**Reply by minad** [reply]
Regarding the column taking too much space, the problem is that Marginalia computes the columns based on the candidate width and not on the display width of the candidate due to Vertico, which Marginalia is not aware of. You could try alternative Marginalia alignment, e.g., `(setq marginalia-align 'center)`. See also https://github.com/jdtsmith/vertico-truncate/issues/2.

**Reply by gety9** [reply]
@minad thank you for reply,

"Marginalia computes the columns based on the candidate"

Could i replace "file candidates" with condidates consisting of custom trancated strings?

I mean if files are:
```
/mydeepdirA/dirA/fileA.el
/mydeepdirB/dirB/fileB.el
...
```

to
```
/dirA/fileA.el
/dirB/fileB.el
...
```

I saw "consult--source-..." customization, but don't think it is that. I think this have to be done on vertico side, right?

**Reply by minad** [reply]
No, it has to be done on the side of Marginalia. There is no good solution for truncation right now, since it affects multiple packages. The affixation-function/annotation-function API is not flexible enough, otherwise we could push the candidate truncation and entire annotation alignment and formatting to the UI, in this case Vertico.

**Reply by gety9** [reply]
got it! thank you for quick replies and amazing work / famility of projects!

**Reply by minad** [reply]
Feel free to keep the discussion open. We hopefully resolve this at some point in the future.
